src/parser.rs

```diff
@@ -0,0 +1,1466 @@
+                l: Some(
+                    list.into_iter()
+                        .map(|x| x.convert_attribute_value())
+                        .collect(),
+                ),
+                ..Default::default()
+            },
+            AttrVal::M(map) => AttributeValue {
+                m: Some(
+                    map.into_iter()
+                        .map(|(key, val)| (key, val.convert_attribute_value()))
+                        .collect(),
+                ),
+                ..Default::default()
+            },
+            AttrVal::NS(list) => AttributeValue {
+                ns: Some(list),
+                ..Default::default()
+            },
+            AttrVal::SS(list) => AttributeValue {
+                ss: Some(list),
+                ..Default::default()
+            },
+            AttrVal::BS(list) => AttributeValue {
+                bs: Some(list),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+impl From<AttrVal> for AttributeValue {
+    fn from(value: AttrVal) -> Self {
+        value.convert_attribute_value()
+    }
+}
+
+/// Parse internal of double quoted string.
+///
+/// It accepts escape characters as the following.
+///
+/// | Escape Sequence | Character Represented by Sequence |
+/// |-----------------|-----------------------------------|
+/// |       \0        | An ASCII NUL (X'00') character    |
+/// |       \r        | A carriage return character       |
+/// |       \n        | A newline (linefeed) character    |
+/// |       \t        | A tab character                   |
+/// |       \\\\      | A backslash (\\) character         |
+/// |       \\\"      | A double quote (") character      |
+/// |       \\\'      | A single quote (') character      |
+fn parse_internal_double_quote_string(str: &str) -> Result<String, EscapeEOS> {
+    let mut result = String::with_capacity(str.len());
+    let mut escaping = false;
+    let mut escaping_pos = 0;
+    for (pos, ch) in str.chars().enumerate() {
+        if escaping {
+            match ch {
+                '0' => result.push('\0'),
+                'r' => result.push('\r'),
+                'n' => result.push('\n'),
+                't' => result.push('\t'),
+                _ => result.push(ch),
+            }
+            escaping = false;
+        } else if ch == '\\' {
+            escaping_pos = pos;
+            escaping = true;
+        } else {
+            result.push(ch);
+        }
+    }
+    if escaping {
+        Err(EscapeEOS {
+            escape_pos: escaping_pos,
+            escape_char: '\\',
+            handling_target: str.to_owned(),
+        })
+    } else {
+        Ok(result)
+    }
+}
+
+/// Parse double quoted string which accepts escape sequence.
+fn parse_double_quote_literal(str: &str) -> Result<String, EscapeEOS> {
+    parse_internal_double_quote_string(&str[1..str.len() - 1])
+}
+
+/// Parse single quoted string as is.
+fn parse_single_quote_literal(str: &str) -> String {
```

Comment:
Would you add test for parse_single_quote_literal just in case?

Reply:
I can do it. But, I'm not sure what you want to protect by test. Could you elaborate on this?

Reply:
From my opinion, the test is to express specification of the source code. So the purpose of the test for parse_single_quote_literal is that we can understand expected behaviour of parse_single_quote_literal that all of char is raw.

Reply:
I added the test into the commit https://github.com/awslabs/dynein/pull/132/commits/8c858c7c722e97f4354e933dba51948235768cc5 to manifest the difference in the handling of escapes.